tools/gdb_pretty_printers/printers.py

```diff
@@ -0,0 +1,432 @@
+
+    def nodemap(self, node):
+        return node.dereference()['impl']['d']['data']['inner']['nodemap']
+
+    def collision_count(self, node):
+        return node.dereference()['impl']['d']['data']['collision']['count']
+
+    def collisions(self, node):
+        return node.dereference()['impl']['d']['data']['collision']['buffer'].address.cast(
+            self.T_ptr)
+
+
+class MapIter(ChampIter):
+    def __init__(self, val):
+        self.T_ptr = gdb.lookup_type("std::pair<" + val.type.template_argument(0).name + ", " +
+                                     val.type.template_argument(1).name + ">").pointer()
+        ChampIter.__init__(self, val)
+        self.pair = None
+
+    def __next__(self):
+        if self.pair:
+            result = ('[%d]' % self.count, self.pair['second'])
+            self.pair = None
+            return result
+        self.pair = super().__next__()
+        return ('[%d]' % self.count, self.pair['first'])
+
+
+class SetIter(ChampIter):
+    def __init__(self, val):
+        self.T_ptr = gdb.lookup_type(val.type.template_argument(0).name).pointer()
+        ChampIter.__init__(self, val)
+
+    def __next__(self):
+        return ('[%d]' % self.count, super().__next__())
+
+
+def num_elements(num):
+    return '1 element' if num == 1 else '%d elements' % num
+
+
+class MapPrinter:
+    "Print an immer::map"
+
+    def __init__(self, val):
+        self.val = val
+
+    def to_string(self):
+        return 'immer::map with %s' % num_elements(self.val['impl_']['size'])
+
+    def children(self):
+        return MapIter(self.val)
+
+    def display_hint(self):
+        return 'map'
+
+
+class SetPrinter:
+    "Prints an immer::set"
+
+    def __init__(self, val):
+        self.val = val
+
+    def to_string(self):
+        return 'immer::set with %s' % num_elements(self.val['impl_']['size'])
+
+    def children(self):
+        return SetIter(self.val)
+
+
+class ListPrinter:
+    "Prints an immer::vector or immer::flex_vector"
+
+    def __init__(self, val, typename):
+        self.val = val
+        self.typename = typename
+
+    def to_string(self):
+        return '%s of length %d' % (self.typename, int(self.val['impl_']['size']))
+
+    def children(self):
+        return ListIter(self.val)
+
+    def display_hint(self):
+        return 'array'
+
+
+def immer_lookup_function(val):
+    compiled_rx = re.compile('^([a-zA-Z0-9_:]+)(<.*>)?$')
+    typename = gdb.types.get_basic_type(val.type).tag
+    if not typename:
+        return None
+    match = compiled_rx.match(typename)
+    if not match:
+        return None
+
+    basename = match.group(1)
+    if basename == "immer::map":
+        return MapPrinter(val)
+    elif basename == "immer::set":
+        return SetPrinter(val)
+    elif basename == "immer::vector":
+        return ListPrinter(val, "immer::vector")
+    elif basename == "immer::flex_vector":
+        return ListPrinter(val, "immer::flex_vector")
```

Comment:
Likewise, maybe `immer::array` with `ListPrinter`

Reply:
The printers in question are dependent on the runtime layout of the type in question, because they essentially have to reimplement the behavior of an iterator in python over the internal structure of the type. I believe array uses a different underlying representation, so a separate iterator would have to be added.

Reply:
It does require a separate iterator, but it's pretty straightforward. I've added an array printer as well.